File: p107.cpp
#include <bits/stdc++.h>
#include <TreeNode.hpp>

using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> result = {{}};
        if(root == NULL) 
            return {};
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});
        int depth = 0;
        while(!q.empty()) {
            pair<TreeNode*, int> curr = q.front();
            if(curr.second > depth) {
                result.insert(result.begin(), {curr.first->val});
                ++depth;
            } else {
                result[0].push_back(curr.first->val);
            }
            if(curr.first->left != NULL)
                q.push({curr.first->left, depth + 1});
            if(curr.first->right != NULL)
                q.push({curr.first->right, depth + 1});
            q.pop();
        }
        return result;
    }
};

File: p107.py
from DS.TreeNode import TreeNode

class Solution:
    def levelOrderBottom(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []

        q = [(root, 0)]
        ans = []
        level = []
        curr_level = 0

        while len(q) > 0:
            curr = q.pop(0)
            if curr[0].left:
                q.append((curr[0].left, curr[1]+1))
            if curr[0].right:
                q.append((curr[0].right, curr[1]+1))
            if curr[1] > curr_level:
                ans.insert(0, level)
                level = []
                curr_level = curr[1]
            level.append(curr[0].val)

        if len(level) > 0:
            ans.insert(0, level)

        return ans
